metaci/api/views/test_testmethod_perf.py

```diff
@@ -0,0 +1,435 @@
+from urllib.parse import urlencode
+import random
+from datetime import timedelta, datetime
+from django.utils import timezone
+
+from rest_framework.test import APIClient, APITestCase
+
+import factory
+import factory.fuzzy
+
+from metaci.plan.models import Plan, PlanRepository
+from metaci.testresults.models import TestResult, TestMethod, TestClass
+from metaci.build.models import BuildFlow, Build
+from metaci.repository.models import Branch, Repository
+
+from metaci.users.models import User
+
+
+class PlanFactory(factory.django.DjangoModelFactory):
+    class Meta:
+        model = Plan
+
+
+class RepositoryFactory(factory.django.DjangoModelFactory):
+    class Meta:
+        model = Repository
+
+    github_id = 1234
+
+
+class PlanRepositoryFactory(factory.django.DjangoModelFactory):
+    class Meta:
+        model = PlanRepository
+
+    plan = factory.SubFactory(PlanFactory)
+    repo = factory.SubFactory(RepositoryFactory)
+
+
+class Branch(factory.django.DjangoModelFactory):
+    class Meta:
+        model = Branch
+
+
+class BuildFactory(factory.django.DjangoModelFactory):
+    class Meta:
+        model = Build
+
+    plan = factory.SubFactory(PlanFactory)
+    repo = factory.SubFactory(RepositoryFactory)
+
+
+class BuildFlowFactory(factory.django.DjangoModelFactory):
+    class Meta:
+        model = BuildFlow
+
+    tests_total = 5
+    build = factory.SubFactory(BuildFactory)
+    flow = "rida"
+
+
+class TestClassFactory(factory.django.DjangoModelFactory):
+    __test__ = False  # PyTest is confused by the classname
+
+    class Meta:
+        model = TestClass
+
+    repo = factory.SubFactory(RepositoryFactory)
+
+
+class TestMethodFactory(factory.django.DjangoModelFactory):
+    __test__ = False  # PyTest is confused by the classname
+
+    class Meta:
+        model = TestMethod
+
+    testclass = factory.SubFactory(TestClassFactory)
+    name = "GenericMethod"
+
+
+class TestResultFactory(factory.django.DjangoModelFactory):
+    __test__ = False  # PyTest is confused by the classname
+
+    class Meta:
+        model = TestResult
+
+    build_flow = factory.SubFactory(BuildFlowFactory)
+    method = factory.SubFactory(TestMethodFactory)
+    duration = 5
+
+
+class UserFactory(factory.django.DjangoModelFactory):
+    class Meta:
+        model = User
+        django_get_or_create = ("username",)
+
+    email = factory.Sequence("user_{}@example.com".format)
+    username = factory.Sequence("user_{}@example.com".format)
+    password = factory.PostGenerationMethodCall("set_password", "foobar")
+    # socialaccount_set = factory.RelatedFactory(SocialAccountFactory, "user")
+
+
+class StaffSuperuserFactory(UserFactory):
+    is_staff = True
+    is_superuser = True
```

Comment:
In the long run, I think we'll want factories for all the models in a central place, rather than duplicating them in each test module that uses them. But we can wait and do that refactoring when we want to use them in a 2nd test module.

Reply:
I agree. I actually did intend to move them but forgot before I checked in. Part of the reason that I didn't get around to it was because it wasn't obvious where to put them. Not sure if there is a convention for file naming for these factories.

We don't have a metaci/tests folder, we don't have a metaci/models folder and I'm not enthusiastic about breaking it up into a bunch of 5-line files.

Reply:
I would create metaci/conftest.py and put them in there. That'll match where they are in MetaDeploy, and is the pytest convention for where to put shared fixtures.

Reply:
https://github.com/SFDO-Tooling/MetaCI/commit/158a21b382ca98e4e5c9422063e026b5bb715902